### DISCUSSION TITLE
Dotnet publish does not export all required runtime libraries

### DISCUSSION THREAD

**Discussion by joriverm** [cefsharp/CefSharp#4486] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

112.2.70

### Operating System

Windows 10

### Architecture

AnyCPU

### .Net Version

.Net 6

### Implementation

WPF

### Reproduction Steps

For a client we are building a WPF application in .net 6 and have switched from WebView2 to CefSharp because WebView2 is rendered in a winforms control, which is *always* on top of the wpf controls even if we use a popup.

Anyway, when doing a dotnet publish of the project, some files are missing just like described in #4022 
However, that issue was for a .net framework project and the solution didn't work + it looks kinda ugly and not ok.
After we imported the CefSharp from nuget, this is what our csproj looked like

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <UseWPF>true</UseWPF>
    <UserSecretsId>bbc8be21-075e-49cb-8003-ea9a60a10467</UserSecretsId>
    <NeutralLanguage>en</NeutralLanguage>
    <CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>
    <ApplicationManifest>app.manifest</ApplicationManifest>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="CefSharp.Wpf.NETCore" Version="112.2.70" />
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
    <PackageReference Include="MinVer" Version="4.3.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Serilog" Version="2.12.0" />
    <PackageReference Include="Serilog.Extensions.Hosting" Version="5.0.1" />
    <PackageReference Include="Serilog.Extensions.Logging" Version="3.1.0" />
    <PackageReference Include="Serilog.Settings.Configuration" Version="3.4.0" />
    <PackageReference Include="Serilog.Sinks.Debug" Version="2.0.0" />
    <PackageReference Include="Serilog.Sinks.File" Version="5.0.0" />
    <PackageReference Include="Serilog.Sinks.Graylog" Version="2.3.0" />
    <PackageReference Include="System.Drawing.Common" Version="7.0.0" />
    <PackageReference Include="Telerik.UI.for.Wpf.60.Xaml" Version="2023.1.315" />
  </ItemGroup>

  <ItemGroup>
	  <Content Include="appsettings.Development.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>Never</CopyToPublishDirectory>
		  <DependentUpon>appsettings.json</DependentUpon>
	  </Content>
	  <Content Include="appsettings.Production.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
		  <DependentUpon>appsettings.json</DependentUpon>
	  </Content>
	  <Content Include="appsettings.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
	  </Content>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="chromiumembeddedframework.runtime.win-arm64" Version="112.2.7" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="chromiumembeddedframework.runtime.win-x64" Version="112.2.7" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="chromiumembeddedframework.runtime.win-x86" Version="112.2.7" />
  </ItemGroup>
</Project>
```

clean as a whistle, as it should be. except im not sure why it added the 3 update package references...
hell, we won't even be needing the arm64 one..
anyway, one publish later and the win-64 runtime of CefSharp is missing in the result.
in a build its there, in a publish it is not.

after changing the PackageReferences for chromiumembeddedframework.runtime.* packages to include, and ignoring the build warning that they are already included, i do get the correct output and all runtime libraries are there.

this makes me believe there is something wrong in the package's configuration/inclusion/publish settings
however, i don't know what as usually all default settings for any package works just fine even for packages i manage myself :/

the final csproj we have now : 
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <UseWPF>true</UseWPF>
    <UserSecretsId>bbc8be21-075e-49cb-8003-ea9a60a10467</UserSecretsId>
    <NeutralLanguage>en</NeutralLanguage>
    <CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>
    <CefSharpExcludeSubProcessExe>true</CefSharpExcludeSubProcessExe>
    <ApplicationManifest>app.manifest</ApplicationManifest>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="CefSharp.Wpf.NETCore" Version="112.3.0" />
    <!-- ¯\_(ツ)_/¯ -->
    <PackageReference Include="chromiumembeddedframework.runtime.win-x64" Version="112.3.0" />
    <PackageReference Include="chromiumembeddedframework.runtime.win-x86" Version="112.3.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
    <PackageReference Include="MinVer" Version="4.3.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Serilog" Version="2.12.0" />
    <PackageReference Include="Serilog.Extensions.Hosting" Version="5.0.1" />
    <PackageReference Include="Serilog.Extensions.Logging" Version="3.1.0" />
    <PackageReference Include="Serilog.Settings.Configuration" Version="3.4.0" />
    <PackageReference Include="Serilog.Sinks.Debug" Version="2.0.0" />
    <PackageReference Include="Serilog.Sinks.File" Version="5.0.0" />
    <PackageReference Include="Serilog.Sinks.Graylog" Version="2.3.0" />
    <PackageReference Include="System.Drawing.Common" Version="7.0.0" />
    <PackageReference Include="Telerik.UI.for.Wpf.60.Xaml" Version="2023.1.315" />
  </ItemGroup>

  <ItemGroup>
	  <Content Include="appsettings.Development.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>Never</CopyToPublishDirectory>
		  <DependentUpon>appsettings.json</DependentUpon>
	  </Content>
	  <Content Include="appsettings.Production.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
		  <DependentUpon>appsettings.json</DependentUpon>
	  </Content>
	  <Content Include="appsettings.json">
		  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
		  <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
		  <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
	  </Content>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="chromiumembeddedframework.runtime.win-arm64" Version="112.2.7" />
  </ItemGroup>
</Project>

```

on top of that, i was diging around and there were a lot of mentions of the MinimalExample repo ( https://github.com/cefsharp/CefSharp.MinimalExample ). however, that one didn't even compile as the functions used in the (imo weird main override) are not found. also, its .netcore csproj is full of things that are not applicable.
As proven by my project here, i don't even think that the Main override is needed?

Anyway, i would love to see this resolved!

Thanks for your time!

### Expected behavior

Runnable, published, package of application

### Actual behavior

application crash on startup because of missing libraries not exported during publish

### Known Workarounds

change generated PackageReferences to include, even if they are already included by the CefSharp.Core.Wpf package

### Does this problem also occur in the CEF Sample Application

Not Tested

**Comment by joriverm** [comment]
thanks for fixing the minimal example. i just did a publish and it has the same issue, libcef and related parts are missing
the cli command i used : `dotnet publish CefSharp.MinimalExample.netcore.sln --output C:\Projects\publish\Example -f net6.0-windows`

**Comment by amaitland** [comment]
Converted to discussion as what you are seeing is fundamentally https://github.com/NuGet/Home/issues/10571 `CefSharp` uses a [runtime.json](https://github.com/cefsharp/cef-binary/blob/master/NuGet/chromiumembeddedframework.runtime.json) file.

When a [RuntimeIdentifier](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) is specified, the build/publish will copy platform specific resources e.g. `win-x64`. Without a [RuntimeIdentifier](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) `nuget/msbuild` won't copy any of the files from the `chromiumembeddedframework.runtime.win-*` packages. 

In development we hack around this by adding the `PackageReference` entries in the [CefSharp.Common.NETCore.targets](https://github.com/cefsharp/CefSharp/blob/cefsharp/112/NuGet/PackageReference/CefSharp.Common.NETCore.targets#L144) file. This works in most cases, not all. 

I always **recommend** setting a `RuntimeIdentifier` during development as per:
- https://github.com/cefsharp/CefSharp/blob/master/NuGet/PackageReference/Readme.txt#L10
- https://github.com/cefsharp/CefSharp/wiki/Quick-Start-For-MS-.Net-5.0-or-greater#20-set-a-runtimeidentifier-in-your-proj-file

> dotnet publish CefSharp.MinimalExample.netcore.sln --output C:\Projects\publish\Example -f net6.0-windows

You can use the new `--use-current-runtime` arg or specify a the explicit runtime identifier.

- https://github.com/cefsharp/CefSharp.MinimalExample#net-core-support
- https://learn.microsoft.com/en-us/dotnet/core/deploying/deploy-with-cli#native-dependencies
- https://learn.microsoft.com/en-us/dotnet/core/rid-catalog#windows-rids

> Architecture
>   AnyCPU

It's my understanding that `.Net Core` family (this include `.Net 6.0`) only supports producing `AnyCPU` **dlls**. For a `AnyCPU WinExe/Exe` you always end up with a `32bit` executable (as that'll run on all three supported platforms). See https://github.com/dotnet/sdk/issues/1906#issuecomment-360863714

I'd encourage you to analyse your exe to see what `arch` it's build for when targeting `AnyCPU`.

You can technically use `dotnet.exe` to launch your applications main entry point dll using an arch specific version or `dotnet.exe`.

I believe publishing multiple platform specific builds of your application is the recommended approach.

**Reply by johnwanzhi** [reply]
Hello, I have a complex problem that has been bothering me for a few days.
Suppose there are two projects A and B in our solution, and the frameworks are both .NET6.0. We use the dotnet method to start the host dll file (in order to support anycpu)
Project A depends on project B, and project B is started through process.start in project A.
Project B OutputType=WinExe refers to the CefSharp.Winforms.NetCore nuget package.

Assuming that project B is anycpu, how to set cefsharp to anycpu?

1>CefSharp After Build Diagnostic
1>CefSharpBuildAction = Default
1>CefSharpExcludeSubProcessExe =
1>RuntimeIdentifier =
1>CefSharpRuntimeIdentifier =
1>Platform = AnyCPU
1>PlatformName = AnyCPU
1>Platforms = AnyCPU
1>PlatformTarget =
1>PlatformTargetAsMSBuildArchitecture = CurrentArchitecture
1>TargetFramework = net6.0-windows
1>TargetFrameworkVersion = v6.0
1>NuGetProjectStyle = PackageReference
1>SelfContained = False
1>PublishSingleFile =
1>OutDir = bin\Debug\net6.0-windows\
1>_CefSharpBrowserSubProcessFiles = C:\Users\johnw\.nuget\packages\cefsharp.common.netcore\109.1.110\runtimes\native\CefSharp.BrowserSubprocess.runtimeconfig.json
1>IncludeAllContentForSelfExtract =
1>IncludeNativeLibrariesForSelfExtract =

**Reply by amaitland** [reply]
> Without a [RuntimeIdentifier](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) nuget/msbuild won't copy any of the files from the chromiumembeddedframework.runtime.win-* packages.

> In development we hack around this by adding the PackageReference entries in the [CefSharp.Common.NETCore.targets](https://github.com/cefsharp/CefSharp/blob/cefsharp/112/NuGet/PackageReference/CefSharp.Common.NETCore.targets#L144) file. This works in most cases, not all.

Did you add package references for the chromiumembeddedframework.runtime.win-* packages?
